### PR TITLE
[WIP] Support empty CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES

### PR DESCRIPTION
**Pull request by eyalroz**
This addresses issue #1975 , or rather - fixes it for when `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` is empty. It seems to me - though I'm not 100% sure - that the other arguments without explicit additions are generated while CMake'ing and are never empty, being relative to the source dir or build dir.

PS:

* Ignore the supposed comments from April (I pushed multiple commits by mistake, then fixed things).
* With this fix introduced, libcudf builds on Devuan Beowulf, which essentially means it builds on Debian Buster (Debian 10).

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -191,8 +191,11 @@ endif(BUILD_TESTS)
 ###################################################################################################
 # - include paths ---------------------------------------------------------------------------------
 
-include_directories("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}"
-                    "${CMAKE_BINARY_DIR}/include"
+if(NOT("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}" STREQUAL ""))
```
Comment (harrism) [inline cpp/CMakeLists.txt]: I think it would be more idiomatic and clearer to just say 
```suggestion
if(CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES)
```

https://cmake.org/cmake/help/v3.0/command/if.html

File: cpp/tests/CMakeLists.txt [hunk 2]
```diff
@@ -35,8 +35,11 @@ endfunction(ConfigureBench)
 ###################################################################################################
 # - include paths ---------------------------------------------------------------------------------
 
-include_directories("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}"
-                    "${CMAKE_BINARY_DIR}/include"
+if(NOT("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}" STREQUAL ""))
```
Comment (harrism) [inline cpp/tests/CMakeLists.txt]:
```suggestion
if(CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES)
```

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -191,8 +191,11 @@ endif(BUILD_TESTS)
 ###################################################################################################
 # - include paths ---------------------------------------------------------------------------------
 
-include_directories("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}"
-                    "${CMAKE_BINARY_DIR}/include"
+if(CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES)
+	include_directories("${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES}")
+endif()
```
Comment (kkraus14) [inline cpp/CMakeLists.txt]: I don't like this because if `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` is empty it can mean that the system doesn't have a usable CUDA installation as opposed to being installed in standard system locations.

From my perspective it would be better to use https://cmake.org/cmake/help/v3.12/module/CheckLanguage.html to check if CUDA can be enabled and then find the relevant directories to include accordingly.

Comment (eyalroz) [reply cpp/CMakeLists.txt]: @kkraus14 : I think you may be mistaken, because we enable CUDA  as a language with the argument to the `project()` command. So the relevant check is probably (not 100% sure) performed already.  So I believe it isn't possible for `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` to be empty except in "legitimate" cases.

Also note that the link says:

>  If `CMAKE_<lang>_COMPILER` is already defined the check does nothing.

And in our case, the CUDA compiler is determined by CMake, i.e. `check_language()` would do nothing in our case IIANM.

caveat: I'm not a CMake guru, so I might be wrong here.

**Comment by harrism** [comment]
@eyalroz can you explain why `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` would ever be empty if the CUDA toolkit is installed?

**Comment by eyalroz** [comment]
> @eyalroz can you explain why `CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES` would ever be empty if the CUDA toolkit is installed?

If the toolkit is installed through a distribution's packages (like it currently is on my system; Devuan Beowulf packages CUDA 9.2), then the CUDA include files are located in `/usr/include/`. And that directory is always part of the C/C++ preprocessor's search path for includes:

```
$ cpp-8 -Wp,-v
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/8/../../../../x86_64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/8/include
 /usr/local/include
 /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
```
see what I mean?

File: cpp/CMakeLists.txt (hunk 3 above)
Comment (eyalroz) [reply cpp/CMakeLists.txt]: @kkraus14 : So, Mark is convinced :-)

Comment (kkraus14) [reply cpp/CMakeLists.txt]: @eyalroz the `project()` command does an `enable_language()` call, not a `check_language()` call. And according to the CMake documentation you can use `check_language()` for support before enabling a language.

Comment (eyalroz) [reply cpp/CMakeLists.txt]: @kkraus14 : Ok, so how about we add:
```
if(not(CMAKE_CUDA_COMPILER))
  message(SEND_ERROR "CMake failed to determine how to compile CUDA")
endif()
```
at the top of our `CMakeLists.txt`'s ?

Comment (kkraus14) [reply cpp/CMakeLists.txt]: > @kkraus14 : Ok, so how about we add:
> 
> ```
> if(not(CMAKE_CUDA_COMPILER))
>   message(SEND_ERROR "CMake failed to determine how to compile CUDA")
> endif()
> ```
> 
> at the top of our `CMakeLists.txt`'s ?

That works, with the caveat that we should reword it to: `CMake failed to find the CUDA compiler`

Comment (eyalroz) [reply cpp/CMakeLists.txt]: @kkraus14 : By the way - for some reason, the CMake our environment uses isn't willing to acknowledge a `check_language()` command. But I'll add the three lines.

**Comment by harrism** [comment]
rerun tests

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -17,6 +17,10 @@ cmake_minimum_required(VERSION 3.12 FATAL_ERROR)
 
 project(CUDA_DATAFRAME VERSION 0.7.2 LANGUAGES C CXX CUDA)
 
+if(NOT CMAKE_CUDA_COMPILER)
+  message(SEND_ERROR "CMake cannot locate a CUDA compiler")
```
Comment (kkraus14) [inline cpp/CMakeLists.txt]: Should we `FATAL_ERROR` this instead?

Comment (eyalroz) [reply cpp/CMakeLists.txt]: No, why? Let's collect all the errors. We don't need the compiler to do Makefile generation.

**Comment by kkraus14** [comment]
@eyalroz Any other changes you intend to make to this PR? Otherwise this looks good to go for me.
